src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -166,101 +182,80 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
+                    sscanf(tmpbuf,"%u",&smoothingGroup);
 			}
 			break;
 
 		case 'u': // usemtl
 			// get name of material
 			{
-				char matName[WORD_BUFFER_LENGTH];
-				bufPtr = goAndCopyNextWord(matName, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
+				bufPtr = goAndCopyNextWord(tmpbuf, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
 #ifdef _IRR_DEBUG_OBJ_LOADER_
-	os::Printer::log("Loaded material start",matName, ELL_DEBUG);
+	os::Printer::log("Loaded material start",tmpbuf, ELL_DEBUG);
 #endif
-				mtlName=matName;
+				mtlName=tmpbuf;
 
                 if (ctx.useMaterials && !ctx.useGroups)
                 {
                     asset::IAsset::E_TYPE types[] {asset::IAsset::ET_SUB_MESH, (asset::IAsset::E_TYPE)0u };
                     auto mb_bundle = _override->findCachedAsset(genKeyForMeshBuf(ctx, _file->getFileName().c_str(), mtlName, grpName), types, ctx.inner, 1u).getContents();
                     if (mb_bundle.first!=mb_bundle.second)
                     {
-                        auto mb = static_cast<asset::ICPUMeshBuffer*>(mb_bundle.first->get());
-                        mb->grab();
-                        SObjMtl* mtl = findMtl(ctx, mtlName, grpName);
-                        ctx.preloadedSubmeshes.insert(std::make_pair(mtl, mb));
+                        submeshes.push_back(*mb_bundle.first);
                     }
-                    else mtlChanged=true;
-
-                    submeshLoadedFromCache = (mb_bundle.first!=mb_bundle.second);
+                    else
+                    {
+                        submeshes.push_back(core::make_smart_refctd_ptr<ICPUMeshBuffer>());
+                        auto found = pipelines.find(mtlName);
+                        assert(found != pipelines.end());
+#ifndef _IRR_DEBUG
+                        if (found != pipelines.end())
+                        {
+#endif
+                            auto& pipeln = found->second;
+                            //cloning pipeline because it will be edited (vertex input params)
+                            //note shallow copy (depth=0), i.e. only pipeline is cloned, but all its sub-assets are taken from original object
+                            submeshes.back()->setPipeline(pipeln.second->clone(0u));
+                            //also make descriptor set
+                            auto metadata = static_cast<const CMTLPipelineMetadata*>(pipeln.second->getMetadata());
+                            images_set_t images = loadImages(pipeln.first.c_str(), metadata->getMaterial());
+                            //intentionally always creating new DS (even if there already is another submesh holding DS with the same descriptors)
+                            //TODO decision to make: try to reuse DSs or leave as it is now? User might want to each submesh's DS in different way, so i'd say leave as is now
+                            auto ds3 = makeDescSet(images, pipeln.second->getLayout()->getDescriptorSetLayout(3u));
```

Comment:
nope cache them, although I'd say you need DS 1 and 3 + PC

1 = VP and VP^T^-1 matrix, etc.
3 = Textures and other material parameters, MVP and other per-draw/per-instance/meshbuffer matrices

Push Constants = small material parameters

I'd try and have the descriptor set reuse approach max, because otherwise we'll get DS swaps before every submesh draw. The eventual UBO I would have as a dynamic one, so the swap of desc set hits the fast path (offset change).

Obviously we acknowledge that the descriptor set 3 itself would be incomplete for rendering and rendering wont work, this is what the pipeline metadata is for, it has some nice REQUIRED INPUTS list for the application to know where to place the camera matrices (which PC bytes, or UBO bytes in a binding+set)

Reply:
there's no way to return DS1, one DS is attached to meshbuffer (as we established, it's 3), but there's no way to return another one

Reply:
also: https://github.com/buildaworldnet/IrrlichtBAW/pull/410#discussion_r362679560

Reply:
You don't return DS1, you leave space for it, the pipeline metadata (this is why I made a base shared class) shall instruct the user how to construct and fill the UBOs used by DS1 + Layout will tell how to make DS1

Reply:
> , this is what the pipeline metadata is for, it has some nice REQUIRED INPUTS list for the application to know where to place the camera matrices (which PC bytes, or UBO bytes in a binding+set)

Reply:
yea, i'll have to adjust CMTLAssetMetadata more to what it's supposed to be